app/code/core/Mage/Adminhtml/controllers/System/AccountController.php

```diff
@@ -102,4 +107,195 @@ public function saveAction()
+                throw new Exception('Missing required fields');
+            }
+
+            // Decode the client data JSON
+            $clientData = json_decode(base64_decode($clientDataJSON), true);
+
+            // Verify challenge
+            $expectedChallenge = $this->_getSession()->getPasskeyChallenge() ?? '';
+            $expectedChallenge = rtrim($expectedChallenge, '=');
+            $clientData['challenge'] = rtrim($clientData['challenge'], '=');
+
+            if (!$expectedChallenge || $clientData['challenge'] !== $expectedChallenge) {
+                throw new Exception('Invalid challenge');
+            }
+
+            // Verify origin
+            $expectedOrigin = Mage::getBaseUrl(Mage_Core_Model_Store::URL_TYPE_WEB);
+            if ($clientData['origin'] !== rtrim($expectedOrigin, '/')) {
+                throw new Exception('Invalid origin');
+            }
+
+            // Check if another user already has this credential
+            $existingUser = Mage::getModel('admin/user')->getCollection()
+                ->addFieldToFilter('passkey_credential_id_hash', $credentialId)
+                ->getFirstItem();
+
+            if ($existingUser->getId() && $existingUser->getId() != $user->getId()) {
+                throw new Exception('Credential already registered to another user');
+            }
+
+            try {
+                // Update the user with the new credential data
+                $user->setPasskeyCredentialIdHash($credentialId)
+                    ->setPasskeyPublicKey($publicKey)
+                    ->save();
+
+                // Clear the challenge from session
+                $this->_getSession()->unsPasskeyChallenge();
+
+                $this->getResponse()->setBodyJson([
+                    'success' => true,
+                    'message' => Mage::helper('adminhtml')->__('Passkey registered successfully'),
+                ]);
+            } catch (Exception $e) {
+                throw new Exception('Failed to save credential: ' . $e->getMessage());
+            }
+        } catch (Exception $e) {
+            $this->getResponse()
+                ->setHttpResponseCode(400)
+                ->setBodyJson([
+                    'success' => false,
+                    'error' => $e->getMessage(),
+                ]);
+        }
+    }
+
+    public function passkeyauthenticateAction()
+    {
+        try {
+            $username = $this->getRequest()->getPost('username');
+            $user = Mage::getModel('admin/user')->loadByUsername($username);
+
+            if (!$user->getId()) {
+                throw new Exception('User not found');
+            }
+
+            $credentials = $user->getResource()->getReadConnection()->fetchAll(
+                "SELECT credential_id_hash FROM {$user->getResource()->getTable('admin/user_credentials')} WHERE user_id = ?",
+                [$user->getId()],
+            );
```

Comment:
couldn't we use a fetchRow or a fetchOne here?

Reply:
Alternatively is there a model we could use `->load($user->getId())` with?

Reply:
actually the user_credentials doesn't even exist, WTH is going on there...

Reply:
It was a leftover method from my first tests